### ISSUE TITLE
Check idmap mounts support before creating a container

### ISSUE THREAD

**Issue by rata** [cri-o/cri-o#7262]
### What happened?

Hi!

I wanted to raise awareness on the fact that is not safe to use idmap mounts in high-level container runtimes without checking the OCI runtime features support. 

The thing is, old runtimes (withot idmap mounts support)  can't fail on unknown fields because the spec [requires to ignore them][spec]. therefore, old OCI runtimes ignore the idmap mounts configs (the new `uidMapping` and `gidMapping` on mounts), which is an issue to use it safely (more on this later). This is sad, but that is how runtimes behave now.

For crun probably this is not an issue here, as you release them in lockstep, right? So CRI-O XX goes with crun YY, and that is always the case. Am I right? In that case it all works, as you can make sure crun YY supports idmap mounts and all is done.

The issue, IIUC, comes with runc support. If CRI-O is used with an old runc that ignore idmap mounts and just start the container, volumes will be mixed up. The UID/GID will be the hostUID/GID of the pod, that is basically garbage, and completely unusable as the hostUID/GID is not guaranteed to be stable over time.

In containerd I'm currently doing the following, to handle this case and give a proper error to users:
 * If userns are in use ==> run "runc features" and check for idmap mounts support.
 * If idmap mounts supports is there, go ahead and create the container. Otherwise, throw a clear error.
 * The container creation might work or not (depends if the host supports idmap mounts for all the fs), but it will never be silently ignored. In the worst case, runc throws an error.

You can check the code I wrote to do this in containerd [here](
https://github.com/containerd/containerd/pull/8287/files#diff-17085223b7307c8c0ed7f2cdcfb904efa5a184a71b186a83961d7feef9e3556fR145-R150).

Exposing info on idmap mounts support is something I did in the runtime-spec due to this issue (not being able to use idmap mounts safely from a high-level container runtime). You can see the PR here: https://github.com/opencontainers/runtime-spec/pull/1219

I've also added support for that in runc and crun:
 * https://github.com/containers/crun/pull/1279
 * https://github.com/opencontainers/runc/pull/3993

Although these changes are not yet included on a release.

Let me know if something is not clear :)

[spec]: https://github.com/opencontainers/runtime-spec/blob/e8c413418efcb816b963be9762ddc202e181ba88/config.md#extensibility

### What did you expect to happen?

---

### How can we reproduce it (as minimally and precisely as possible)?

---

### CRI-O and Kubernetes version

<details>

```console
$ crio --version
# paste output here
```

```console
$ kubectl version --output=json
# paste output here
```

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here
```

</details>

### Additional environment details (AWS, VirtualBox, physical, etc.)

<details>

</details>

**Comment by haircommander** [comment]
Thanks for making us aware @rata ! Do you have any bandwidth/desire to add a `$runtime features` check? For whoever does it: we have a set of packages in `internal/config/` that do static checks like this. We could add a new one: `internal/config/ociruntime` where we go through all of the configured runtimes and initialize their support of various features (wasi could be another good one cc @sohankunkerkar )  

> For crun probably this is not an issue here, as you release them in lockstep, right? 

nope! we release independently :)

**Comment by AdityaVardhanSingh** [comment]
Hey, can I work on this?

**Comment by rata** [comment]
@haircommander awesome! @AdityaVardhanSingh sure, you can do it! Let me know if you can't, I can try to find some time to work on this :)

**Comment by MeenuyD** [comment]
Hello, @rata I would like to work on this issue can you guide me on how to get started to resolve?

**Comment by rata** [comment]
@MeenuyD it seems @AdityaVardhanSingh will work on this already

**Comment by MeenuyD** [comment]
ok

**Comment by AdityaVardhanSingh** [comment]
Hey, @rata I think I got the issue but what are we expecting here to be done?

**Comment by haircommander** [comment]
I actually would suggest an alternative approach. Instead of re-checking whether runc/crun support the feature each time a container with idmap is created, I would use a similar mechanism to the `internal/config/node/` package. When the config is generated in `pkg/config/config.go` we have a list of the runtime handlers. Once that's finalized, you could loop through the handlers and call `$runtime features` on them, and save whether they support idmap as part of the handler object. Then, in container creation, you could just query that field

**Comment by rata** [comment]
@haircommander that seems way better! :)

**Comment by AdityaVardhanSingh** [comment]
@rata can we connect?? I have somethings to clarify.

**Comment by rata** [comment]
@AdityaVardhanSingh what do you mean? You can post here. @haircommander will probably have better suggestions, he is a maintainer and I've never checkout the cri-o repo :-)

**Comment by haircommander** [comment]
You can reach out to me on k8s slack if you'd like to chat elsewhere @AdityaVardhanSingh :)

**Comment by AdityaVardhanSingh** [comment]
That would be really helpful.

**Comment by AdityaVardhanSingh** [comment]
@rata I am working on this but many things came up which was new to me so just trying to do it in the best possible way. Just wanted to give my best, if you can give me a couple of days more if that is ok?

**Comment by rata** [comment]
We have blog posts scheduled for next week that depend on this, so I'm sorry, but I don't think we can :(

btw, crun 1.9 is out now, it ships with the PR I linked in the description, to expose idmap mounts support.

**Comment by rata** [comment]
This is fixed in main now, backported to 1.28 and will be included in the next path release.

Thanks all for your help!

@AdityaVardhanSingh I hope you find another good first issue to collaborate. Thanks a lot anyways! :)
